### ISSUE TITLE
x/tools/gopls: DocumentHighlight for format strings

### ISSUE THREAD

**Issue by findleyr** [golang/go#70050]
This just occurred to me, when looking up a rarely used format verb for the nth time at pkg.go.dev/fmt: ~we should offer hover information for format strings. To start, this could simply show documentation for the active format verb under the cursor, but we could also extend it to explain more complicated formatting expressions.~ (EDIT: DocumentHighlight is a more useful feature; repurposing this issue for that)

We could extract the tricky bits of detecting and parsing format strings from the printf analyzer, though that begs the question of why we can't detect printf wrappers...  @adonovan I wonder what you think about this idea, and more generally about the concept of exposing facts from the analysis framework to other gopls functionality.

Note that accurate detection of printf wrappers is not strictly necessary--that's more of a thought experiment. For hover, we could use a much weaker heuristic to guess whether a string is likely to be a format string.

**Comment by adonovan** [comment]
Format strings are only one small step removed from the language spec in the sense that every Go file uses them and every Go reader is expected to recognize them, or at least the dozen or so most common ones. Just as I would find it odd to serve a hover box over the `go` keyword that says it starts a new goroutine, it seems to me redundant to show a dialog over %q saying it quotes a string. Though it may benefit the beginner, I expect it would do so at the cost of the Go programmers who have passed that phase and find the extra popups a nuisance.

To your bigger implied questions:

- Yes, I'm sure there's a lot we could do with better integration, though the mechanics seem tricky: hover depends only on typed syntax, whereas facts depend on analysis. Perhaps, just as analyses can report SuggestedFixes, which we tie together with diagnostics, they could also [waves hands] contribute information to hover reports.

- A somewhat related issue for interpreting "little languages" in string literals such as fmt, SQL, and text/template (and perhaps third-party ones like https://github.com/a-h/templ if that is the de facto victor), is https://github.com/golang/go/issues/64543. I think templ has a VS Code extension for completions, but I assume that only works in standalone files. If raw string literals are commonly used, it would be neat if we could parse the strings, interpret their nonstandard types (as we do with fmt), and show both diagnostics and cross-references in a general way, at the correct portion of the string literal.

**Comment by findleyr** [comment]
> Though it may benefit the beginner, I expect it would do so at the cost of the Go programmers who have passed that phase and find the extra popups a nuisance.

I'm not following this argument. FWIW, this issue is premised on the fact that this would help _me_, and I'm not a beginner! We offer hover for builtins, even though they're part of the spec... Certainly one uses `make` more than `%p`. You may argue that 'make' depends on its arguments, but so does the behavior of `%p`!  (And for the record, I never have to look up `%q`, even though its behavior is rather more complicated than others).  In fact, I'd go so far as to say it would be nice to offer completion after typing `%`.

Unfortunately, you may be right that the hover could be a nuisance, but then aren't all hovers a nuisance? In my editor, I only trigger hover explicitly.

**Comment by adonovan** [comment]
Fair point. I remember I once asked you why we bothered with hover on built-ins and you responded that they're really useful.

I do think there would be real value in tying each %v formatting operation with its operand, as when they are numerous it can be hard to see the correspondence at a glance (which is why Python now has f-strings). Perhaps DocumentHighlight is the ideal LSP UI? I still have concerns about efficiency if lightweight queries such as Hover and Highlight were to invoke the analysis framework, even if only in particular cases with a restricted subset of analyzers. We hack up a little experiment.

**Comment by xzbdmw** [comment]
> I'd go so far as to say it would be nice to offer completion after typing %

It would be so much better if
```go
var foo int
fmt.Printf("%foo")
```
can expands to 
```go
var foo int
fmt.Printf("%v", foo)
```
via applyEdits with completion **and** code action, given that go has rejected f-string proposal, it is always a pain if arguments are too many.

**Comment by xzbdmw** [comment]
> Perhaps DocumentHighlight is the ideal LSP UI? I still have concerns about efficiency if lightweight queries such as Hover and Highlight were to invoke the analysis framework, even if only in particular cases with a restricted subset of analyzers. We hack up a little experiment.

DocumentHighlight should solve the "counting" process quite well, I'm writing a lot long printf so I want to implement this feature, where should I start?

**Comment by adonovan** [comment]
> DocumentHighlight should solve the "counting" process quite well, I'm writing a lot long printf so I want to implement this feature, where should I start?

You'll need a function that, given a string literal and an offset, returns an index within the denoted string; and a function that, given a logical format string and an offset, returns the index of the % conversion (if any) that it falls within. The simplest way to do the second one would be to follow the approach in https://github.com/golang/go/issues/68279#issuecomment-2282907630, which isn't perfect, but is good enough for >99.9% of strings.

Once you have those, hooking into internal/golang/highlight.go for CallExpr shouldn't be hard. The only tricky bit is how to know which functions to apply it to: obviously fmt.Printf, but what about all the wrappers? We could use the analysis framework, but there are a lot of tricky problems there. Since this is only highlighting, it needn't be perfect: perhaps it's enough to observe that the format string contains a %, or approximately as many %s as arguments?

**Comment by xzbdmw** [comment]
I think this can be simpler without parsing format string and args. The format syntax is `%[flags][width][.precision][verb]`, what we need for DocumentHighlight is only the range of the placeholder, that is, a range starts with '%' and ends with a valid verb, and details sits between them can be ignored.

**Comment by xzbdmw** [comment]
Should we also paint the format string with semantic highlight? Once we have a iter describing each element of a string in order, highlighting becomes free too, though vscode's textmate rule and [treesitter-printf](https://github.com/pstuifzand/tree-sitter-printf) can detect something similar.

**Comment by xzbdmw** [comment]
https://github.com/golang/go/issues/68279#issuecomment-2282907630, This method has limitations that can't handle explicit argument indexes, eg. `fmt.Sprintf("%[2]d %[1]d\n", 11, 22)`, also directives of `%T`, `%p`, will not hit [wrapper logic](https://github.com/golang/go/blob/c46ba1f9efeb8efa6d8f513ec92f36ae1bfb8cd0/src/fmt/print.go#L756), because https://github.com/golang/go/blob/c46ba1f9efeb8efa6d8f513ec92f36ae1bfb8cd0/src/fmt/print.go#L695-L704

Is these acceptable or should I go another way?

**Comment by adonovan** [comment]
Good point about %T and %p; sorry I didn't notice that sooner.

> Is these acceptable or should I go another way?

Feel free to take another approach, such as writing a parser for the format string (with a suitable test suite---perhaps you can devise a test where your parser is cross-checked against the actual implementation of fmt.Printf to ensure that it computes the correct answer).

**Comment by xzbdmw** [comment]
Can we do this
`fmt.Sprintf("%[3]*.[2]*[1]f, foo: %v", 12.0, 2, 6, fooVar)` hover -> `fmt.Sprintf("%6.2{12.0}, foo: {fooVar}")`

**Comment by findleyr** [comment]
It sounds like there is not a lot of interest in hover for format strings, so let's repurpose this issue to be just about DocumentHighlight, and close. Hover support is certainly not happening for v0.18.0.

Thanks @xzbdmw for this very nice feature.
